Add shelly ble scanner support

bdraco opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add shelly ble scanner support

Requires shelly beta firmware 0.12.0-beta2 (update via the web ui)

PR Splits

Waiting for review
- [x] `BaseRemoteHaScanner` - https://github.com/home-assistant/core/pull/82012 
- [x] model splits https://github.com/home-assistant-libs/home-assistant-bluetooth/pull/10
- [x] move `decode.py` into Bluetooth data tools https://github.com/home-assistant/core/pull/82016 
- [x] new pr for `async_subscribe_events` on the coordinator https://github.com/home-assistant/core/pull/82052
- [x] move any part of `bluetooth/__init__.py` into `aioshelly` that can be moved https://github.com/home-assistant-libs/aioshelly/pull/281
- [x] json performance https://github.com/home-assistant-libs/aioshelly/pull/282
- [x] Add on connected and on disconnected callbacks to start/stop/unregister the scanner https://github.com/home-assistant-libs/aioshelly/pull/283

Changes

- [x] Config flow option for disable/active/passive scanning

Tests

- [x] options flow


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/24945

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdraco committed:
Add shelly ble scanner support

bdraco committed:
init

bdraco committed:
write a decoder

bdraco committed:
write a decoder

bdraco committed:
hook it in

bdraco committed:
fixes

bdraco committed:
handle bad data

bdraco committed:
fix address compat

bdraco committed:
fix address compat

bdraco committed:
fix address compat

bdraco committed:
script upload

bdraco committed:
script upload

bdraco committed:
script upload

bdraco commented on homeassistant/components/shelly/coordinator.py:
```diff
@@ -378,6 +380,19 @@ def update_sleep_period(self) -> bool:
 
         return True
 
+    @callback
+    def async_subscribe_ble_events(
+        self, ble_callback: Callable[[dict[str, Any]], None]
+    ) -> CALLBACK_TYPE:
+        """Subscribe to BLE events."""
+
+        def _unsubscribe() -> None:
+            self._ble_callback = None
+
+        self._ble_callback = ble_callback
+
+        return _unsubscribe
+
```
Maybe make this a more generic subscriber for events

bdraco commented on homeassistant/components/shelly/coordinator.py:
```diff
@@ -404,6 +419,8 @@ def _async_device_updates_handler(self) -> None:
                     ENTRY_RELOAD_COOLDOWN,
                 )
                 self.hass.async_create_task(self._debounced_reload.async_call())
+            elif self._ble_callback and event_type == BLE_SCAN_RESULT_EVENT:
+                self._ble_callback(event)
```
Maybe make this a more generic subscriber for events

bdraco committed:
tweaks

bdraco committed:
fixes

bdraco committed:
fix setup race

bdraco committed:
active switch

bdraco committed:
reduce

bdraco committed:
refactoring

bdraco committed:
fix merge

bdraco committed:
tweak

bdraco committed:
byte order

bdraco committed:
tweaks

bdraco committed:
tweak

bdraco committed:
Move remote bluetooth remote scanner implementation into a base class

- The same implementation can be reused for other remotes
  such as shelly in #82007

bdraco committed:
Merge branch 'remote_scanner_base_class' into shelly_ble

bdraco committed:
upstream some code

bdraco committed:
deps

bdraco committed:
naming

bdraco committed:
Merge branch 'remote_scanner_base_class' into shelly_ble

bdraco committed:
naming

bdraco committed:
Merge branch 'dev' into remote_scanner_base_class

bdraco committed:
tests

bdraco committed:
coverage

bdraco committed:
Merge branch 'remote_scanner_base_class' into shelly_ble

bdraco committed:
merge

bdraco committed:
Merge branch 'remote_scanner_base_class' into shelly_ble

kzyapkov commented on homeassistant/components/shelly/bluetooth/__init__.py:
```diff
@@ -0,0 +1,97 @@
+BLE.Scanner.Subscribe(function (ev, res) {
+    if (ev === BLE.Scanner.SCAN_RESULT) {
+        Shelly.emitEvent("ble.scan_result", [
+            res.addr,
```
Maybe put a leading entry here for a "format version", in case it needs to change later

bdraco replied:
Will probably make the version the ha version so we know where it came from

Will also make the event formatted as the same const

bdraco replied:
b67a146ac574c104817548786eeb2968be445fe4

thecode commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -249,7 +249,13 @@ def _async_rpc_device_setup() -> None:
+
+        # TODO: only connect the scanner if enabled in config options active/true/false
+        if not sleep_period:
+            entry.async_on_unload(
```
I think we can move this bellow line 250

bdraco replied:
Will likely need to tweak it a bit since we only want to turn it on if the shelly is up and running since we need to pull the scanner if we loose the connection to the shelly so bluetooth knows it has gone away and will immediately switch to the next nearest one for a given device.

For that we need to know when it gets disconnected.  I couldn't find a way to get callbacks when the connection is established or disconnected so I think that probably will have to be a change in `aioshelly`

thecode replied:
Actually if we reached this stage the device is already connected, else we will get an error raised here:
https://github.com/home-assistant/core/blob/f17041e239f243324801076a814a4d487aede946/homeassistant/components/shelly/__init__.py#L276

We had a disconnect event which I removed since it was not used, I can add it back. We still know if device is disconnected if it doesn't send an event within a time period:
https://github.com/home-assistant/core/blob/f17041e239f243324801076a814a4d487aede946/homeassistant/components/shelly/coordinator.py#L450

bdraco replied:
> We had a disconnect event which I removed since it was not used, I can add it back. We still know if device is disconnected if it doesn't send an event within a time period:

That would be great.

The goal is to cancel the scanner when we get disconnected via the callback returned from `async_connect_scanner`

https://github.com/home-assistant/core/pull/82007/files#diff-ef0eaf3053c447ec8f2cbfd32a151528d6eb6a9d012d10a9e86e73e706c7bbe5R47

...and than when we reconnect, connect the scanner back up by calling `async_connect_scanner` again

bdraco committed:
Merge remote-tracking branch 'upstream/dev' into shelly_ble

bdraco committed:
tweaks

bdraco committed:
const

bdraco commented on homeassistant/components/shelly/bluetooth/__init__.py:
```diff
@@ -0,0 +1,93 @@
+from .scanner import ShellyBLEScanner
+
+
+class ShellyScript(TypedDict, total=False):
+    """Shelly Script."""
+
+    id: int
+    name: str
+    enable: bool
+    running: bool
+
+
+async def _async_get_scripts_by_name(device: RpcDevice) -> dict[str, int]:
+    """Get scripts by name."""
+    data = await device.call_rpc("Script.List")
+    scripts: list[ShellyScript] = data["scripts"]
+    return {script["name"]: script["id"] for script in scripts}
+
```
Is this something we should move into `aioshelly`?

thecode replied:
Yes

bdraco committed:
version guard

bdraco committed:
Merge branch 'dev' into shelly_ble

bdraco committed:
Add the ability to register for shelly event callbacks

This is an internal change only to support #82007

bdraco committed:
refactoring to prepare to move to lib

bdraco committed:
break out more

bdraco committed:
break out more

bdraco committed:
move code to aioshelly

bdraco committed:
drop requires

bdraco committed:
Avoid writing shelly state unless status changes

We always wrote the state via async_set_updated_data when
there was an event. With many BLE events this results
in hundreds of state writes per minute

bdraco committed:
Avoid writing shelly state unless status changes

We always wrote the state via async_set_updated_data when
there was an event. With many BLE events this results
in hundreds of state writes per minute

bdraco committed:
Merge branch 'shelly_event_callbacks' into shelly_ble

bdraco committed:
fix compat

bdraco committed:
fix mocking in test

bdraco committed:
Merge branch 'shelly_event_callbacks' into shelly_ble

bdraco committed:
Merge branch 'dev' into shelly_ble

bdraco commented on homeassistant/components/shelly/__init__.py:
```diff
@@ -249,7 +251,17 @@ def _async_rpc_device_setup() -> None:
+        # TODO: only connect the scanner if enabled in config options active/true/false
+        if (
+            not sleep_period
+            and device.shelly
```
```suggestion
```

bdraco committed:
Update homeassistant/components/shelly/__init__.py

bdraco committed:
bump

bdraco committed:
Merge remote-tracking branch 'upstream/shelly_ble' into shelly_ble

bdraco committed:
move to coordinator

bdraco committed:
circular import

bdraco committed:
handle initialized before subscribed

bdraco committed:
clear callbacks on disconnected

bdraco committed:
tweak name

bdraco committed:
bump lib

bdraco committed:
add config option for ble scanner mode

bdraco committed:
reload when we change scanning

bdraco committed:
reconfigure when we change scanning

bdraco committed:
reconfigure when we change scanning

bdraco committed:
reconfigure when we change scanning

bdraco committed:
scanner mode option

bdraco committed:
tweak

bdraco committed:
fixes

bdraco committed:
fix sleepy

bdraco committed:
temp pickup codeowner for this cycle so I get notifies

bdraco committed:
options flow tests

bdraco committed:
exclude scanner

bdraco committed:
relax version check

bdraco committed:
Merge branch 'dev' into shelly_ble

balloob commented on homeassistant/components/shelly/coordinator.py:
```diff
@@ -460,21 +476,71 @@ def sw_version(self) -> str:
-        event = device.event
-        status = device.status
+        if AwesomeVersion(self.device.version) < BLE_MIN_VERSION:
+            LOGGER.error(
```
For a future PR we can make this a repair issue.

bdraco replied:
Should probably do the same for esphome < 2022.11.x

balloob commented on homeassistant/components/shelly/config_flow.py:
```diff
@@ -37,6 +43,13 @@
 
+BLE_SCANNER_OPTIONS = [
+    selector.SelectOptionDict(value=BLEScannerMode.DISABLED, label="Disabled"),
+    selector.SelectOptionDict(value=BLEScannerMode.ACTIVE, label="Active"),
```
I wish we could make it more clear what it means to be active/passive.

bdraco replied:
<img width="574" alt="Screenshot 2022-11-15 at 7 49 09 AM" src="https://user-images.githubusercontent.com/663432/201935831-203465b2-b6b6-48ba-a8a0-6e4834209c91.png">

Added a description

balloob commented on homeassistant/components/shelly/config_flow.py:
```diff
@@ -310,3 +323,50 @@ async def _async_get_info(self, host: str) -> dict[str, Any]:
+        if user_input is not None:
+            return self.async_create_entry(title="", data=user_input)
+
+        return self.async_show_form(
```
Should we check the version here and tell the user it should update it's Shelly first?

bdraco replied:
will now abort if they choose active or passive and the firmware is too old

thecode commented on .coveragerc:
```diff
@@ -1110,6 +1110,7 @@ omit =
     homeassistant/components/seven_segments/image_processing.py
     homeassistant/components/seventeentrack/sensor.py
     homeassistant/components/shelly/binary_sensor.py
+    homeassistant/components/shelly/bluetooth/scanner.py
```
Can we add tests to cover this file? I am trying to add tests coverage for Shelly since it is the only part missing to qualify for platinum scale.

bdraco replied:
done

bdraco committed:
abort if ble is unsupported and tell them why so they can figure it out

bdraco committed:
active vs passive

bdraco committed:
Merge branch 'dev' into shelly_ble

bdraco committed:
fix py39

bdraco committed:
scanner tests

bdraco committed:
scanner tests

bdraco committed:
py3.9 again

bdraco committed:
py3.9 again

bdraco committed:
py3.9

bdraco committed:
one more

bdraco committed:
Merge branch 'dev' into shelly_ble

thecode reviewed (approved):
Great works, thanks!
